socorro/lib/transaction.py

```diff
@@ -8,161 +8,107 @@
-                # these exceptions may or may not be retriable the test is
-                # for is a last ditch effort to see if we can retry
-                if not self.db_conn_context_source.is_operational_exception(x):
-                    # If the logger exists, log the issue, otherwise print it
-                    # to stdout.
-                    if self.config.logger:
-                        self.config.logger.critical(
-                            'Unrecoverable %s transaction error',
-                            self.connection_source_type,
-                            exc_info=True
-                        )
-                    else:
-                        print('Unrecoverable %s transaction error' % self.connection_source_type)
-                    reraise(*last_failure)
-
-                self.config.logger.critical(
-                    '%s transaction error eligible for retry',
-                    self.connection_source_type,
-                    exc_info=True
-                )
+                conn.rollback()
+            except Exception:
+                # NOTE(willkg): This assumes the connection context has a configured
+                # logger; if that ever changes, we need to fix this
+                conn.config.logger.exception('cannot rollback')
+            six.reraise(*excinfo)
+
+
+# Backoff amounts in seconds; last is a 0 so it doesn't sleep after the last
+# failure before exiting the loop
+BACKOFF_TIMES = [1, 1, 1, 1, 1, 0]
+
+
+def retry(connection_context, quit_check, fun, **kwargs):
+    """Runs the function retrying periodically
 
-            except self.db_conn_context_source.operational_exceptions:
-                self.config.logger.critical(
-                    '%s transaction error eligible for retry',
-                    self.connection_source_type,
+    The connection context must define the following:
+
+    * ``is_retryable_exception(Exception) -> bool``
+
+      If this returns true, then ``retry` will call ``force_reconnect()``,
+      wait some time, and try again.
+
+    * ``force_reconnect() -> None``
+
+      Forces recreation of the connection the connection context wraps.
+
+    * ``config.logger.critical``
+
+      The connection context should have a configman DotDict for configuration
+      with a logger that implements ``warning``.
+
+    :arg connection_context: the connection context
+    :arg quit_check: the quit_check function
+    :arg fun: the function to retry
+    :arg kwargs: named arguments to pass to the function
+
+    :returns: varies
+
+    """
+    last_failure = None
+
+    for backoff_time in BACKOFF_TIMES:
+        try:
+            with connection_context() as conn:
+                return fun(conn, **kwargs)
+        except Exception as exc:
+            last_failure = sys.exc_info()
+            if connection_context.is_retryable_exception(exc):
+                connection_context.config.logger.warning(
```

Comment:
This happens all the time in the processor when it's processing a crash it hasn't processed before which is usually the case, so this is kind of annoying.

I think I want to change this so it only logs on the last exception thrown rather than the interim exceptions.

Reply:
On second thought, I think I'm going to nix the logging altogether. The caller can log the error if it wants to.